ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,567 @@
+    return expirationTime < referenceTime && expirationTime != Utils.Infinite_Time;
+  }
+
+  private void processDeleteIndexForSegment(IndexValue putIndexValue, ScanResults results, String key,
+      Long referenceTime) {
+    if (putIndexValue.getExpiresAtMs() == -1) {
+      results.updateSegmentBaseMap(key, putIndexValue.getSize() * -1);
+    } else if (!isExpired(putIndexValue.getExpiresAtMs(), referenceTime)) {
+      results.updateSegmentBaseMap(key, putIndexValue.getSize() * -1);
+      if (results.checkIfBucketEntryExists(putIndexValue.getExpiresAtMs(), key)) {
+        results.updateSegmentBucket(putIndexValue.getExpiresAtMs(), key, putIndexValue.getSize());
+      }
+    }
+  }
+
+  private void processDeleteIndexForContainer(IndexValue putIndexValue, ScanResults results, String serviceId,
+      String containerId, Long referenceTime) {
+    if (putIndexValue.getExpiresAtMs() == -1) {
+      results.updateContainerBaseMap(serviceId, containerId, putIndexValue.getSize() * -1);
+    } else if (!isExpired(putIndexValue.getExpiresAtMs(), referenceTime)) {
+      results.updateContainerBaseMap(serviceId, containerId, putIndexValue.getSize() * -1);
+      if (results.checkIfBucketEntryExists(putIndexValue.getExpiresAtMs(), serviceId, containerId)) {
+        results.updateContainerBucket(putIndexValue.getExpiresAtMs(), serviceId, containerId,
+            putIndexValue.getSize());
+      }
+    }
+  }
+
+  private boolean isWithinWindow(long referenceTime, long window, long offSet) {
+    return referenceTime < window - offSet;
+  }
+
+  private class IndexScanner implements Runnable {
+    final long scanTimeInMs;
+    final long scanTimeWithOffsetInMs;
+    final ScanResults newScanResults;
+    final long scanWindowBoundary;
+    final long requestedScanReferenceTime;
+
+    IndexScanner(long scanTimeInMs, long requestedScanReferenceTime) {
+      this.scanTimeInMs = scanTimeInMs;
+      this.scanTimeWithOffsetInMs = scanTimeInMs - segmentScanOffset;
+      this.newScanResults = new ScanResults(bucketCount, bucketTimeSpan);
+      this.scanWindowBoundary = newScanResults.createBuckets(scanTimeInMs, segmentScanOffset);
+      this.requestedScanReferenceTime = requestedScanReferenceTime;
+    }
+
+    @Override
+    public void run() {
+      Offset checkpoint = prepareForScanning();
+      Offset previousSegmentOffset = null;
+      try {
+        for (Map.Entry<Offset, IndexSegment> indexSegmentEntry : index.indexes.entrySet()) {
+          IndexSegment segment = indexSegmentEntry.getValue();
+          if (segment.getStartOffset().compareTo(checkpoint) >= 0) {
+            break;
+          }
+          List<MessageInfo> messageInfos = new ArrayList<>();
+          segment.getEntriesSince(null, new FindEntriesCondition(Integer.MAX_VALUE),
+              messageInfos, new AtomicLong(0));
+          FileSpan currentSpan = new FileSpan(segment.getStartOffset(), segment.getStartOffset());
```

Comment:
Yes, I only want to search in a single index segment and the existing implementation of findKey uses subMap(start, end) inclusive to determine the search space.